lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -104,23 +105,121 @@ class PatchEntryPointMutate : public Patch {
+    // calls must be kept.
+    bool pushConstSpill = false;
+    // Per-push-const-offset lists of loads from push const. We attempt to unspill these.
+    SmallVector<UserDataNodeUsage, 8> pushConstOffsets;
+    // Per-user-data-offset lists of lgc.root.descriptor calls
+    SmallVector<UserDataNodeUsage, 8> rootDescriptors;
+    // Per-descriptor-set lists of lgc.descriptor.set calls
+    SmallVector<UserDataNodeUsage, 8> descriptorSets;
+    // Minimum offset at which spill table is used.
+    unsigned spillUsage = UINT_MAX;
+  };
+
+  // Gather user data usage in all shaders.
+  void gatherUserDataUsage(llvm::Module *module);
+
+  // Fix up user data uses.
+  void fixupUserDataUses(Module &module);
+
   void processShader();
 
-  FunctionType *generateEntryPointType(uint64_t *inRegMask) const;
+  FunctionType *generateEntryPointType(uint64_t *inRegMask);
 
   void addSpecialUserDataArgs(SmallVectorImpl<UserDataArg> &userDataArgs,
                               SmallVectorImpl<UserDataArg> &specialUserDataArgs, IRBuilder<> &builder) const;
-  void addUserDataArgs(SmallVectorImpl<UserDataArg> &userDataArgs, IRBuilder<> &builder) const;
+  void addUserDataArgs(SmallVectorImpl<UserDataArg> &userDataArgs, IRBuilder<> &builder);
+  unsigned addUserDataArg(SmallVectorImpl<UserDataArg> &userDataArgs, unsigned userDataValue, unsigned sizeInDwords,
+                          unsigned *argIndex, bool useFixedLayout, unsigned userDataSize, IRBuilder<> &builder);
+
   void determineUnspilledUserDataArgs(ArrayRef<UserDataArg> userDataArgs, ArrayRef<UserDataArg> specialUserDataArgs,
-                                      IRBuilder<> &builder, SmallVectorImpl<UserDataArg> &unspilledArgs) const;
+                                      IRBuilder<> &builder, SmallVectorImpl<UserDataArg> &unspilledArgs);
 
   uint64_t pushFixedShaderArgTys(SmallVectorImpl<Type *> &argTys) const;
 
-  bool isResourceNodeActive(const ResourceNode *node, bool isRootNode) const;
+  // Get UserDataUsage struct for the merged shader stage that contains the given shader stage
+  UserDataUsage *getUserDataUsage(ShaderStage stage);
+
+  // Get the shader stage that the given shader stage is merged into.
+  ShaderStage getMergedShaderStage(ShaderStage stage) const;
 
   bool m_hasTs;                             // Whether the pipeline has tessllation shader
   bool m_hasGs;                             // Whether the pipeline has geometry shader
   PipelineState *m_pipelineState = nullptr; // Pipeline state from PipelineStateWrapper pass
+  // Per-HW-shader-stage gathered user data usage information.
+  llvm::SmallVector<std::unique_ptr<UserDataUsage>, ShaderStageCount> m_userDataUsage;
+};
+
+// =====================================================================================================================
+// Utility class to generate code to extend an i32 address into a 64-bit pointer, using either PC or a given
+// constant high half value.
+class PointerExtender {
```

Comment:
I think the pointer is used as address, maybe rename it to AddressExtender or something similar. I just believe pointer is too general.